cmd/dra-example-webhook/main.go

```diff
@@ -0,0 +1,304 @@
+	if err != nil {
+		return nil, fmt.Errorf("request could not be decoded: %w", err)
+	}
+
+	if *gvk != admissionv1.SchemeGroupVersion.WithKind("AdmissionReview") {
+		return nil, fmt.Errorf("unsupported group version kind: %v", gvk)
+	}
+
+	requestedAdmissionReview, ok := obj.(*admissionv1.AdmissionReview)
+	if !ok {
+		return nil, fmt.Errorf("expected v1.AdmissionReview but got: %T", obj)
+	}
+
+	return requestedAdmissionReview, nil
+}
+
+// admitResourceClaimParameters accepts both ResourceClaims and ResourceClaimTemplates and validates their
+// opaque device configuration parameters for this driver.
+func admitResourceClaimParameters(ar admissionv1.AdmissionReview) *admissionv1.AdmissionResponse {
+	klog.V(2).Info("admitting resource claim parameters")
+
+	var deviceConfigs []resourceapi.DeviceClaimConfiguration
+	var specPath string
+
+	raw := ar.Request.Object.Raw
+	deserializer := codecs.UniversalDeserializer()
+
+	switch ar.Request.Resource {
+	case resourceClaimResource:
+		claim := resourceapi.ResourceClaim{}
+		if _, _, err := deserializer.Decode(raw, nil, &claim); err != nil {
+			klog.Error(err)
+			return &admissionv1.AdmissionResponse{
+				Result: &metav1.Status{
+					Message: err.Error(),
+					Reason:  metav1.StatusReasonBadRequest,
+				},
+			}
+		}
+		deviceConfigs = claim.Spec.Devices.Config
+		specPath = "spec"
+	case resourceClaimTemplateResource:
+		claimTemplate := resourceapi.ResourceClaimTemplate{}
+		if _, _, err := deserializer.Decode(raw, nil, &claimTemplate); err != nil {
+			klog.Error(err)
+			return &admissionv1.AdmissionResponse{
+				Result: &metav1.Status{
+					Message: err.Error(),
+					Reason:  metav1.StatusReasonBadRequest,
+				},
+			}
+		}
+		deviceConfigs = claimTemplate.Spec.Spec.Devices.Config
+		specPath = "spec.spec"
+	default:
+		msg := fmt.Sprintf("expect resource to be %s or %s, got %s", resourceClaimResource, resourceClaimTemplateResource, ar.Request.Resource)
+		klog.Error(msg)
+		return &admissionv1.AdmissionResponse{
+			Result: &metav1.Status{
+				Message: msg,
+				Reason:  metav1.StatusReasonBadRequest,
+			},
+		}
+	}
+
+	var errs []error
+	for configIndex, config := range deviceConfigs {
+		if config.Opaque == nil || config.Opaque.Driver != consts.DriverName {
+			continue
+		}
+
+		fieldPath := fmt.Sprintf("%s.devices.config[%d].opaque.parameters", specPath, configIndex)
+		decodedConfig, err := runtime.Decode(configapi.Decoder, config.DeviceConfiguration.Opaque.Parameters.Raw)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("error decoding object at %s: %w", fieldPath, err))
+			continue
+		}
+		gpuConfig, ok := decodedConfig.(*configapi.GpuConfig)
+		if !ok {
+			errs = append(errs, fmt.Errorf("expected v1alpha1.GpuConfig at %s but got: %T", fieldPath, decodedConfig))
+			continue
+		}
+		err = gpuConfig.Validate()
+		if err != nil {
+			errs = append(errs, fmt.Errorf("object at %s is invalid: %w", fieldPath, err))
+		}
```

Comment:
Since this is about the extent of the driver-specific logic, I wonder if we could eventually make most of this a library in https://github.com/kubernetes/dynamic-resource-allocation where driver authors would only have to provide this custom type, a validation function, and an invocation from a `main` package. Getting some mileage on this implementation though seems wise before making it too easy :)